kernel/syscall.c

```diff
@@ -1785,6 +1793,41 @@ long myst_syscall_symlink(const char* target, const char* linkpath)
     return ret;
 }
 
+long myst_syscall_symlinkat(
+    const char* target,
+    int newdirfd,
+    const char* linkpath)
+{
+    long ret = 0;
+    struct locals
+    {
+        char abspath[PATH_MAX];
+    };
+    struct locals* locals = NULL;
+
+    /* If linkpath is absolute, then ignore dirfd */
+    if (*linkpath == '/' || newdirfd == AT_FDCWD)
+    {
+        ret = myst_syscall_symlink(target, linkpath);
```

Comment:
This pattern is repeated enough times, we check should `dirfd` be ignored or not and call `myst_get_absolute_path_from_dirfd` is not. I wonder if we can consolidate and simply the logic as:

```c
myst_get_absolute_path_from_dirfd(dirfd, char* path_in, char** path_out)
{
    if (dirfd == AT_FDCWD || *path_in == '/')
    {
        *path_out = path_in;
    }
   else 
  {
      *path_out = malloc(....)
       ... resolve and copy the path_out ...
  }
}
```
And then in the caller's clean up code, it determines if there is heap variable needs to free with:
```c
  if (*path_out != path_in)
      free(*path_out);
```

Reply:
[resolve_at_path()](https://github.com/deislabs/mystikos/blob/main/kernel/chown.c#L204) in kernel/chown.c does something similar. Maybe that can be reused here.

Reply:
@jxyang Good idea! But there seems to be a small issue that `pathname` (`path_in` in your sample code) is of type `const char*`. Is type conversion OK per our standard?

Reply:
> resolve_at_path

@vtikoo Thanks for pointing it out. Do you mean to reuse `resolve_at_path` in `myst_get_absolute_path_from_dirfd`? Or to reuse `myst_get_absolute_path_from_dirfd` in `resolve_at_path`? Or to merge them into one function?

Reply:
`resolve_at_path` is at a higher level than `myst_get_absolute_path_from_dirfd`. You can check `myst_syscall_fchownat` in kernel/chown.c as a reference usage of `resolve_at_path`.

Reply:
Oh, I got what you mean. Would you mind I move `resolve_at_path` into syscall.c?

Reply:
Sure, go for it 👍

Reply:
Note `resolve_at_path` expects the caller to allocate memory, while my proposed `myst_get_absolute_path_from_dirfd` expects the callee to allocate memory (because it has better idea about how long the resolved path should be).

@Francis-Liu  You can cast away the `const` qualifier.